// angular2/es6/dev/src/render/dom/convert.js
import { ListWrapper, MapWrapper } from 'angular2/src/facade/collection';
import { isPresent, isArray } from 'angular2/src/facade/lang';
import { DirectiveMetadata } from 'angular2/src/render/api';
/**
 * Converts a [DirectiveMetadata] to a map representation. This creates a copy,
 * that is, subsequent changes to `meta` will not be mirrored in the map.
 */
export function directiveMetadataToMap(meta) {
    return MapWrapper.createFromPairs([
        ['id', meta.id],
        ['selector', meta.selector],
        ['compileChildren', meta.compileChildren],
        ['hostProperties', _cloneIfPresent(meta.hostProperties)],
        ['hostListeners', _cloneIfPresent(meta.hostListeners)],
        ['hostActions', _cloneIfPresent(meta.hostActions)],
        ['hostAttributes', _cloneIfPresent(meta.hostAttributes)],
        ['properties', _cloneIfPresent(meta.properties)],
        ['readAttributes', _cloneIfPresent(meta.readAttributes)],
        ['type', meta.type],
        ['exportAs', meta.exportAs],
        ['callOnDestroy', meta.callOnDestroy],
        ['callOnCheck', meta.callOnCheck],
        ['callOnInit', meta.callOnInit],
        ['callOnChange', meta.callOnChange],
        ['callOnAllChangesDone', meta.callOnAllChangesDone],
        ['events', meta.events],
        ['changeDetection', meta.changeDetection],
        ['version', 1],
    ]);
}
/**
 * Converts a map representation of [DirectiveMetadata] into a
 * [DirectiveMetadata] object. This creates a copy, that is, subsequent changes
 * to `map` will not be mirrored in the [DirectiveMetadata] object.
 */
export function directiveMetadataFromMap(map) {
    return new DirectiveMetadata({
        id: MapWrapper.get(map, 'id'),
        selector: MapWrapper.get(map, 'selector'),
        compileChildren: MapWrapper.get(map, 'compileChildren'),
        hostProperties: _cloneIfPresent(MapWrapper.get(map, 'hostProperties')),
        hostListeners: _cloneIfPresent(MapWrapper.get(map, 'hostListeners')),
        hostActions: _cloneIfPresent(MapWrapper.get(map, 'hostActions')),
        hostAttributes: _cloneIfPresent(MapWrapper.get(map, 'hostAttributes')),
        properties: _cloneIfPresent(MapWrapper.get(map, 'properties')),
        readAttributes: _cloneIfPresent(MapWrapper.get(map, 'readAttributes')),
        type: MapWrapper.get(map, 'type'),
        exportAs: MapWrapper.get(map, 'exportAs'),
        callOnDestroy: MapWrapper.get(map, 'callOnDestroy'),
        callOnCheck: MapWrapper.get(map, 'callOnCheck'),
        callOnChange: MapWrapper.get(map, 'callOnChange'),
        callOnInit: MapWrapper.get(map, 'callOnInit'),
        callOnAllChangesDone: MapWrapper.get(map, 'callOnAllChangesDone'),
        events: _cloneIfPresent(MapWrapper.get(map, 'events')),
        changeDetection: MapWrapper.get(map, 'changeDetection'),
    });
}
/**
 * Clones the [List] or [Map] `o` if it is present.
 */
function _cloneIfPresent(o) {
    if (!isPresent(o))
        return null;
    return isArray(o) ? ListWrapper.clone(o) : MapWrapper.clone(o);
}
