// angular2/es6/prod/src/forms/directives/ng_model.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") return Reflect.decorate(decorators, target, key, desc);
    switch (arguments.length) {
        case 2: return decorators.reduceRight(function(o, d) { return (d && d(o)) || o; }, target);
        case 3: return decorators.reduceRight(function(o, d) { return (d && d(target, key)), void 0; }, void 0);
        case 4: return decorators.reduceRight(function(o, d) { return (d && d(target, key, o)) || o; }, desc);
    }
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { CONST_EXPR } from 'angular2/src/facade/lang';
import { EventEmitter, ObservableWrapper } from 'angular2/src/facade/async';
import { StringMapWrapper } from 'angular2/src/facade/collection';
import { Directive, onChange } from 'angular2/angular2';
import { forwardRef, Binding } from 'angular2/di';
import { NgControl } from './ng_control';
import { Control } from '../model';
import { setUpControl } from './shared';
const formControlBinding = CONST_EXPR(new Binding(NgControl, { toAlias: forwardRef(() => NgModel) }));
/**
 * Binds a domain model to the form.
 *
 * # Example
 *  ```
 * @Component({selector: "search-comp"})
 * @View({
 *      directives: [formDirectives],
 *      template: `
              <input type='text' [(ng-model)]="searchQuery">
 *      `})
 * class SearchComp {
 *  searchQuery: string;
 * }
 *  ```
 *
 * @exportedAs angular2/forms
 */
export let NgModel = class extends NgControl {
    constructor(...args) {
        super(...args);
        this._control = new Control("");
        this._added = false;
        this.ngModel = new EventEmitter();
    }
    onChange(c) {
        if (!this._added) {
            setUpControl(this._control, this);
            this.control.updateValidity();
            this._added = true;
        }
        ;
        if (StringMapWrapper.contains(c, "model")) {
            this._control.updateValue(this.model);
        }
    }
    get control() { return this._control; }
    get path() { return []; }
    viewToModelUpdate(newValue) { ObservableWrapper.callNext(this.ngModel, newValue); }
};
NgModel = __decorate([
    Directive({
        selector: '[ng-model]:not([ng-control]):not([ng-form-control])',
        hostInjector: [formControlBinding],
        properties: ['model: ng-model'],
        events: ['ngModel'],
        lifecycle: [onChange],
        exportAs: 'form'
    }), 
    __metadata('design:paramtypes', [])
], NgModel);
