// angular2/es6/prod/src/render/dom/view/proto_view.js
import { isPresent } from 'angular2/src/facade/lang';
import { DOM } from 'angular2/src/dom/dom_adapter';
import { ListWrapper } from 'angular2/src/facade/collection';
import { NG_BINDING_CLASS } from '../util';
import { RenderProtoViewRef } from '../../api';
export function resolveInternalDomProtoView(protoViewRef) {
    return protoViewRef._protoView;
}
export class DomProtoViewRef extends RenderProtoViewRef {
    constructor(_protoView) {
        super();
        this._protoView = _protoView;
    }
}
export class DomProtoView {
    constructor({ elementBinders, element, transitiveContentTagCount }) {
        this.element = element;
        this.elementBinders = elementBinders;
        this.transitiveContentTagCount = transitiveContentTagCount;
        this.isTemplateElement = DOM.isTemplateElement(this.element);
        this.rootBindingOffset =
            (isPresent(this.element) && DOM.hasClass(this.element, NG_BINDING_CLASS)) ? 1 : 0;
        this.boundTextNodeCount =
            ListWrapper.reduce(elementBinders, (prevCount, elementBinder) => prevCount + elementBinder.textNodeIndices.length, 0);
        this.rootNodeCount =
            this.isTemplateElement ? DOM.childNodes(DOM.content(this.element)).length : 1;
    }
}
