// angular2/es6/prod/src/render/dom/view/view.js
import { DOM } from 'angular2/src/dom/dom_adapter';
import { MapWrapper } from 'angular2/src/facade/collection';
import { Locals } from 'angular2/change_detection';
import { isPresent } from 'angular2/src/facade/lang';
import { RenderViewRef } from '../../api';
export function resolveInternalDomView(viewRef) {
    return viewRef._view;
}
export class DomViewRef extends RenderViewRef {
    constructor(_view) {
        super();
        this._view = _view;
    }
}
/**
 * Const of making objects: http://jsperf.com/instantiate-size-of-object
 */
export class DomView {
    constructor(proto, rootNodes, boundTextNodes, boundElements) {
        this.proto = proto;
        this.rootNodes = rootNodes;
        this.boundTextNodes = boundTextNodes;
        this.boundElements = boundElements;
        this.hostLightDom = null;
        this.shadowRoot = null;
        this.hydrated = false;
        this.eventDispatcher = null;
        this.eventHandlerRemovers = [];
    }
    getDirectParentElement(boundElementIndex) {
        var binder = this.proto.elementBinders[boundElementIndex];
        var parent = null;
        if (binder.parentIndex !== -1 && binder.distanceToParent === 1) {
            parent = this.boundElements[binder.parentIndex];
        }
        return parent;
    }
    setElementProperty(elementIndex, propertyName, value) {
        var setter = MapWrapper.get(this.proto.elementBinders[elementIndex].propertySetters, propertyName);
        setter(this.boundElements[elementIndex].element, value);
    }
    callAction(elementIndex, actionExpression, actionArgs) {
        var binder = this.proto.elementBinders[elementIndex];
        var hostAction = MapWrapper.get(binder.hostActions, actionExpression);
        hostAction.eval(this.boundElements[elementIndex].element, this._localsWithAction(actionArgs));
    }
    _localsWithAction(action) {
        var map = MapWrapper.create();
        MapWrapper.set(map, '$action', action);
        return new Locals(null, map);
    }
    setText(textIndex, value) { DOM.setText(this.boundTextNodes[textIndex], value); }
    dispatchEvent(elementIndex, eventName, event) {
        var allowDefaultBehavior = true;
        if (isPresent(this.eventDispatcher)) {
            var evalLocals = MapWrapper.create();
            MapWrapper.set(evalLocals, '$event', event);
            // TODO(tbosch): reenable this when we are parsing element properties
            // out of action expressions
            // var localValues = this.proto.elementBinders[elementIndex].eventLocals.eval(null, new
            // Locals(null, evalLocals));
            // this.eventDispatcher.dispatchEvent(elementIndex, eventName, localValues);
            allowDefaultBehavior =
                this.eventDispatcher.dispatchEvent(elementIndex, eventName, evalLocals);
            if (!allowDefaultBehavior) {
                event.preventDefault();
            }
        }
        return allowDefaultBehavior;
    }
}
