// angular2/src/http/url_search_params.js
var lang_1 = require('angular2/src/facade/lang');
var collection_1 = require('angular2/src/facade/collection');
function paramParser(rawParams) {
    var map = collection_1.MapWrapper.create();
    var params = lang_1.StringWrapper.split(rawParams, '&');
    collection_1.ListWrapper.forEach(params, function (param) {
        var split = lang_1.StringWrapper.split(param, '=');
        var key = collection_1.ListWrapper.get(split, 0);
        var val = collection_1.ListWrapper.get(split, 1);
        var list = collection_1.MapWrapper.get(map, key) || collection_1.ListWrapper.create();
        collection_1.ListWrapper.push(list, val);
        collection_1.MapWrapper.set(map, key, list);
    });
    return map;
}
var URLSearchParams = (function () {
    function URLSearchParams(rawParams) {
        this.rawParams = rawParams;
        this.paramsMap = paramParser(rawParams);
    }
    URLSearchParams.prototype.has = function (param) { return collection_1.MapWrapper.contains(this.paramsMap, param); };
    URLSearchParams.prototype.get = function (param) { return collection_1.ListWrapper.first(collection_1.MapWrapper.get(this.paramsMap, param)); };
    URLSearchParams.prototype.getAll = function (param) {
        return collection_1.MapWrapper.get(this.paramsMap, param) || collection_1.ListWrapper.create();
    };
    URLSearchParams.prototype.append = function (param, val) {
        var list = collection_1.MapWrapper.get(this.paramsMap, param) || collection_1.ListWrapper.create();
        collection_1.ListWrapper.push(list, val);
        collection_1.MapWrapper.set(this.paramsMap, param, list);
    };
    URLSearchParams.prototype.toString = function () {
        var paramsList = collection_1.ListWrapper.create();
        collection_1.MapWrapper.forEach(this.paramsMap, function (values, k) {
            collection_1.ListWrapper.forEach(values, function (v) { collection_1.ListWrapper.push(paramsList, k + '=' + v); });
        });
        return collection_1.ListWrapper.join(paramsList, '&');
    };
    URLSearchParams.prototype.delete = function (param) { collection_1.MapWrapper.delete(this.paramsMap, param); };
    return URLSearchParams;
})();
exports.URLSearchParams = URLSearchParams;
exports.__esModule = true;
